guppylang/compiler/expr_compiler.py

```diff
@@ -452,68 +463,104 @@ def visit_ResultExpr(self, node: ResultExpr) -> Wire:
-                return
-
-            # Otherwise, compile the first iterator and construct a TailLoop
-            gen, *gens = gens
-            compiler.compile_stmts([gen.iter_assign], self.dfg)
-            assert isinstance(gen.iter, PlaceNode)
-            assert isinstance(gen.hasnext, PlaceNode)
-            inputs = [gen.iter, list_name]
-            with self._new_loop(inputs, gen.hasnext):
-                # If there is a next element, compile it and continue with the next
-                # generator
+        def build_update(elt_port: Wire) -> None:
+            """Callback for pushing a comprehension element onto the list."""
+            list_port = self.dfg[list_place]
+            [], [self.dfg[list_place]] = self._build_method_call(
+                list_ty, "append", node, [list_port, elt_port], list_ty.args
+            )
+
+        self._compile_generators(node.elt, node.generators, [list_place], build_update)
+        return self.dfg[list_place]
+
+    def visit_DesugaredArrayComp(self, node: DesugaredArrayComp) -> Wire:
+        # Allocate an uninitialised array of the desired size and a counter variable
+        array_ty = get_type(node)
+        assert isinstance(array_ty, OpaqueType)
+        array_var = Variable(next(tmp_vars), array_ty, node)
+        count_var = Variable(next(tmp_vars), int_type(), node)
+        hugr_elt_ty = (
+            ht.Option(node.elt_ty.to_hugr())
+            if node.elt_ty.linear
+            else node.elt_ty.to_hugr()
+        )
+        self.dfg[array_var] = self.builder.add_op(
+            array_new_uninitialized(hugr_elt_ty, node.length)
+        )
+        self.dfg[count_var] = self.builder.load(
+            hugr.std.int.IntVal(0, width=NumericType.INT_WIDTH)
+        )
+
+        def build_update(elt: Wire) -> None:
+            """Callback for putting an element into the array."""
+            array, count = self.dfg[array_var], self.dfg[count_var]
+            [], [self.dfg[array_var]] = self._build_method_call(
+                array_ty, "__setitem__", node, [array, count, elt], array_ty.args
+            )
+            # Update `count += 1`
+            one = self.builder.load(hugr.std.int.IntVal(1, width=NumericType.INT_WIDTH))
+            [self.dfg[count_var]], [] = self._build_method_call(
+                int_type(), "__add__", node, [count, one]
+            )
+
+        self._compile_generators(
+            node.elt, [node.generator], [array_var, count_var], build_update
+        )
+        return self.dfg[array_var]
+
+    def _build_method_call(
+        self,
+        ty: Type,
+        method: str,
+        node: AstNode,
+        args: list[Wire],
+        type_args: Sequence[Argument] | None = None,
+    ) -> CallReturnWires:
+        func = self.globals.get_instance_func(ty, method)
+        assert func is not None
+        return func.compile_call(args, type_args or [], self.dfg, self.globals, node)
+
+    def _compile_generators(
```

Comment:
Ok, I was just wondering whether you could use `ExitStack` to make the whole `_compile_generators` nonrecursive too....and you'd gone and done it :-D :-) Good job! 👍 

Next...I am wondering if you can turn the whole `_compile_generators` into a context manager. The `build_update` callback is a bit restrictive (lots of hidden state because `build_update` has a fixed single-Wire interface, etc.), so rather than `self._compile_generators(node.elt, node.generators, [list_place], build_update)`, you could call it by
```
with _compile_generators(self, node.generators, [list_place]) as _:
  build_update(node.elt)
```
....which would allow inlining `build_update` at the use site.

For the implementation of `_compile_generators`, see the [`@contextmanager`](https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager) decorator - you'd put your `yield` in place of the existing call to the `build_update` Callable.

Reply:
Love the idea! 👍